File: internal/limiter/limiter.go
package limiter

import (
	"context"
	"sync"
	"time"
)

type Limiter struct {
	mu sync.RWMutex
	buckets map[string]*tokenBucket
	capacity float64
	refillRate float64
	idleTTL time.Duration
}


// Create a new limiter. Each key has a separate bucket.
func New(capacity int, window time.Duration, idleTTL time.Duration) *Limiter {
	return &Limiter{
		buckets: make(map[string]*tokenBucket),
		capacity: float64(capacity),
		refillRate: float64(capacity) / window.Seconds(),
		idleTTL: idleTTL,	
	}
}


// Check bucket for key
func (l *Limiter) Allow(key string, cost float64) (allowed bool, remaining int, limit int, retryAfter time.Duration) {
	b := l.getOrCreateBucket(key)
	ok, rem, retry := b.allow(cost)
	return ok, int(rem), int(l.capacity), retry
}


func (l *Limiter) getOrCreateBucket(key string) *tokenBucket {
	l.mu.RLock()
	b, ok := l.buckets[key]
	l.mu.RUnlock()
	if ok {
		return b
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	if b, ok = l.buckets[key]; ok {
		return b
	}

	b = newTokenBucket(l.capacity, l.refillRate)
	l.buckets[key] = b
	return b
}


func (l *Limiter) StartCleanup(ctx context.Context, interval time.Duration) {
	go func() {
		ticket := time.NewTicker(interval)
		defer ticket.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticket.C:
				l.evictIdle()
			}
		}
	}()
}

func (l *Limiter) evictIdle() {
	cutoff := time.Now().Add(-l.idleTTL)

	l.mu.RLock()
	stale := make([]string, 0)
	for key, b := range l.buckets {
		if b.idleSince(cutoff) {
			stale = append(stale, key)
		}
	}
	l.mu.RUnlock()

	if len(stale) == 0 {
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()
	for _, key := range stale {
		if b, ok := l.buckets[key]; ok && b.idleSince(cutoff) {
			delete(l.buckets, key)
		}
	}
}

func (l *Limiter) Len() int {
	l.mu.RLock()
	defer l.mu.RUnlock()
	return len(l.buckets)
}

File: app/limiter.py
"""
Limiter manages one TokenBucket per client, plus a background task to clean up idle buckets.
Simplified from Go's version thanks to the use of Python's dicts.
"""

from __future__ import annotations
import asyncio
from dataclasses import dataclass
import time

from app.bucket import AllowResult, TokenBucket

@dataclass
class CheckResult:
    allowed: bool
    remaining: int
    limit: int
    retry_after_seconds: float

class Limiter:
    def __init__(self, capacity: int, window_seconds: float, idle_ttl_seconds: float) -> None:
        self._buckets: dict[str, TokenBucket] = {}
        self._creation_lock = asyncio.Lock()
        self.capacity = float(capacity)
        self.refill_rate = capacity / window_seconds
        self.idle_ttl_seconds = idle_ttl_seconds
        self._cleanup_task = asyncio.Task | None = None

    async def _get_or_create_bucket(self, key: str) -> TokenBucket:
        bucket = self._buckets.get(key)
        if bucket is not None:
            return bucket
        
        # Lock only around the check and create to avoid two coroutines racing to create the same bucket.
        async with self._creation_lock:
            bucket = self._buckets.get(key)
            if bucket is None:
                bucket = TokenBucket(self.capacity, self.refill_rate)
                self._buckets[key] = bucket
            return bucket

    async def allow(self, key: str, cost: float = 1.0) -> CheckResult:
        bucket = await self._get_or_create_bucket(key)
        result: AllowResult = await bucket.allow(cost)
        return CheckResult(
            allowed=result.allowed,
            remaining=int(result.remaining),
            limit=int(self.capacity),
            retry_after_seconds=result.retry_after_seconds,
        )
    
    async def evict_idle(self) -> None:
        cutoff = time.monotonic() - self.idle_ttl_seconds

        stale: list[str] = []
        for key, bucket in self._buckets.items():
            if await bucket.idle_since(cutoff):
                stale.append(key)

        async with self._creation_lock:
            for key in stale:
                bucket = self._buckets.get(key)
                if bucket is not None and await bucket.idle_since(cutoff):
                    del self._buckets[key]

    async def _cleanup_loop(self, interval_seconds: float) -> None:
        try:
            while True:
                await asyncio.sleep(interval_seconds)
                await self.evict_idle()
        except asyncio.CancelledError:
            pass

    def start_cleanup(self, interval_seconds: float) -> None:
        """Launch the background eviction loop as an asyncio task."""
        self._cleanup_task = asyncio.create_task(self._cleanup_loop(interval_seconds))

    async def stop_cleanup(self) -> None:
        if self._cleanup_task is not None:
            self._cleanup_task.cancel()
            await asyncio.gather(self._cleanup_task, return_exceptions=True)

    def __len__(self) -> int:
        return len(self._buckets)